src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1759,3 +1761,38 @@ MultiSubqueryPushdownTable(Query *subquery)
 
 	return subqueryTableNode;
 }
+
+
+/*
+ * PartitionColumnForPushedDownSubquery finds the partition column on the target
+ * list of a pushed down subquery.
+ */
+static Var *
+PartitionColumnForPushedDownSubquery(Query *query)
+{
+	List *targetEntryList = query->targetList;
+
+	TargetEntry *targetEntry = NULL;
+	foreach_ptr(targetEntry, targetEntryList)
+	{
+		Expr *targetExpression = targetEntry->expr;
+		if (IsA(targetExpression, Var))
+		{
+			bool isPartitionColumn = IsPartitionColumn(targetExpression, query);
+			if (isPartitionColumn)
```

Comment:
The main question that I have: Is it always enough to find the first column? Could we have any cases where because of the order  of the target list we consistently pushdown or not.

I think the answer is yes, but still considering any potential edge cases.

Reply:
> The main question that I have: Is it always enough to find the first column?

Not always, but often enough. 

For instance, if we're trying to determine whether a group by is safe to push down, finding any partition column on the target list would be sufficient, since any of those columns would have to be in the group by as well. Hence, these are all equally pushdownable.

```sql
select x1, count(*) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a group by x1;
select x2, count(*) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a group by x2;
select x1, x2, count(*) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a group by x1, x2;
```

An exception is:
```sql
select count(distinct x1), count(distinct x2) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a;
```

which would require use to mark both columns as partition column

Conversely, the following are both pushed down:
```sql
select count(distinct x2) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a;
select count(distinct x2) from (select t1.x x1, t2.x x2 from test t1 left join test t2 on (t1.x = t2.x) group by t1.x, t2.x) a;
```
because x1 does not end up on the target list of the MultiTable in the second case

INSERT..SELECT does its own set of checks for partition columns, so it does not care much which of the 2 partition columns goes into the destination distribution column.